File: 2nd-try/199/199.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<int> rightSideView(TreeNode *root)
    {
        vector<int> res;

        if (!root)
            return res;

        queue<TreeNode *> qt;
        qt.push(root);
        while (!empty(qt))
        {
            int curLevelSize = qt.size();
            for (int i = 0; i < curLevelSize - 1; i++){
                if (qt.front()->left)
                    qt.push(qt.front()->left);
                if (qt.front()->right)
                    qt.push(qt.front()->right);
                qt.pop();
            }
            if (qt.front()->left)
                qt.push(qt.front()->left);
            if (qt.front()->right)
                qt.push(qt.front()->right);
            res.push_back(qt.front()->val);
            qt.pop();
        }
        return res;
    }
};

File: 2nd-try/binary_tree/levelOrderTraversal/go/LevelOrder/199.go
package LevelOrder

func RightSideView(root *TreeNode) []int {
	// 思路：层序遍历，只把每层最右节点元素加入结果集即可
	res := []int{}
	if root == nil {
		return res
	}
	q := Queue[*TreeNode]{}
	q.Push(root)
	for !q.Empty() {
		size := q.Size()
		for i := 0; i < size-1; i++ {
			if q.Front().Left != nil {
				q.Push(q.Front().Left)
			}
			if q.Front().Right != nil {
				q.Push(q.Front().Right)
			}
			q.Pop()
		}
		// 每层的最后一个节点
		res = append(res, q.Front().Val)
		if q.Front().Left != nil {
			q.Push(q.Front().Left)
		}
		if q.Front().Right != nil {
			q.Push(q.Front().Right)
		}
		q.Pop()
	}
	return res
}
